File: stacks/api-first/fastapi-app/schemas/order.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from schemas.user import OrderUserRead


OrderStatus = Literal["created", "paid", "shipped", "cancelled"]


class OrderItemCreate(BaseModel):
    product_id: int
    quantity: int = Field(gt=0)


class OrderCreate(BaseModel):
    user_id: int
    items: list[OrderItemCreate] = Field(min_length=1)


class OrderStatusUpdate(BaseModel):
    status: OrderStatus


class OrderItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    product_id: int
    product_name: str
    quantity: int
    unit_price: float


class OrderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user: OrderUserRead
    items: list[OrderItemRead]
    item_count: int
    total: float
    status: OrderStatus
    created_at: datetime

File: stacks/api-first/express-app/src/serializers/order.js
function serializeOrder(order) {
  return {
    id: order.id,
    user: {
      id: order.user.id,
      name: order.user.name
    },
    items: order.items.map((item) => ({
      product_id: item.product_id,
      product_name: item.product.name,
      quantity: item.quantity,
      unit_price: item.unit_price
    })),
    item_count: order.items.length,
    total: order.total,
    status: order.status,
    created_at: order.created_at
  };
}

module.exports = {
  serializeOrder
};
